File: src/tools/suggest.ts
import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { ZenMoneyAPI } from "../api.js";
import type { ZenState } from "../state.js";

export function registerSuggestTools(
  server: McpServer,
  api: ZenMoneyAPI,
  state: ZenState
) {
  server.tool(
    "suggest_category",
    "Get ZenMoney's auto-suggestion for category and merchant based on a payee name. Useful for categorizing receipts.",
    {
      payee: z.string().describe("The payee/merchant name from the receipt"),
    },
    async ({ payee }) => {
      try {
        // Best effort: state is only used to turn ids into names, so a failed
        // sync should not block the suggestion itself.
        await state.ensureSynced().catch(() => {});

        const results = await api.suggest([{ payee }]);
        const suggestion = results[0];

        if (!suggestion) {
          return {
            content: [
              {
                type: "text" as const,
                text: `No suggestions found for "${payee}".`,
              },
            ],
          };
        }

        const parts: string[] = [`Suggestions for "${payee}":`];

        if (suggestion.tag && suggestion.tag.length > 0) {
          const tagNames = suggestion.tag
            .map((tagId) => {
              const tag = state.tags.find((t) => t.id === tagId);
              return tag ? `${tag.title} (\`${tag.id}\`)` : `\`${tagId}\``;
            })
            .join(", ");
          parts.push(`- Category: ${tagNames}`);
        }

        if (suggestion.merchant) {
          const merchant = state.merchants.find(
            (m) => m.id === suggestion.merchant
          );
          const name = merchant ? merchant.title : suggestion.merchant;
          parts.push(`- Merchant: ${name}`);
        }

        if (suggestion.payee) {
          parts.push(`- Normalized payee: ${suggestion.payee}`);
        }

        return {
          content: [{ type: "text" as const, text: parts.join("\n") }],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: `Suggest failed: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );
}

File: go/internal/tools/suggest.go
package tools

import (
	"context"
	"strings"

	"github.com/modelcontextprotocol/go-sdk/mcp"

	"github.com/david-dvinskykh/zenmoney-mcp-ddvin/go/internal/zen"
)

type suggestArgs struct {
	Payee string `json:"payee" jsonschema:"The payee/merchant name from the receipt"`
}

func registerSuggest(server *mcp.Server, api *zen.API, st *zen.State) {
	mcp.AddTool(server, &mcp.Tool{
		Name: "suggest_category",
		Description: "Get ZenMoney's auto-suggestion for category and merchant based on a payee name. " +
			"Useful for categorizing receipts.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, args suggestArgs) (*mcp.CallToolResult, any, error) {
		// Best effort: state is only used to turn ids into names, so a failed
		// sync should not block the suggestion itself.
		_ = st.EnsureSynced(ctx)

		results, err := api.Suggest(ctx, []zen.SuggestRequest{{Payee: args.Payee}})
		if err != nil {
			return errorResult("Suggest failed: " + err.Error()), nil, nil
		}
		if len(results) == 0 {
			return textResult(`No suggestions found for "` + args.Payee + `".`), nil, nil
		}

		s := results[0]
		parts := []string{`Suggestions for "` + args.Payee + `":`}

		if len(s.Tag) > 0 {
			names := make([]string, 0, len(s.Tag))
			for _, id := range s.Tag {
				if tag := findTag(st, id); tag != nil {
					names = append(names, tag.Title+" (`"+tag.ID+"`)")
				} else {
					names = append(names, "`"+id+"`")
				}
			}
			parts = append(parts, "- Category: "+strings.Join(names, ", "))
		}

		if s.Merchant != nil {
			name := *s.Merchant
			if m := findMerchant(st, *s.Merchant); m != nil {
				name = m.Title
			}
			parts = append(parts, "- Merchant: "+name)
		}

		if s.Payee != nil {
			parts = append(parts, "- Normalized payee: "+*s.Payee)
		}

		return textResult(strings.Join(parts, "\n")), nil, nil
	})
}
